tsfc/node.py

```diff
@@ -98,7 +98,8 @@ def get_hash(self):
 def traversal(expression_dags):
     """Pre-order traversal of the nodes of expression DAGs."""
     seen = set(expression_dags)
-    lifo = list(expression_dags)
+    # Some roots might be same, but they must be visited only once
+    lifo = list(seen)
```

Comment:
Careful, `lifo` might come out in a different order on different executions.  Does that matter?